### PR TITLE
Support nested property expansion and compaction.

### PR DESCRIPTION
**Pull request by gkellogg**
Note, `_expandObject` is mostly a simple re-factor of the body of `expand`.

Processing of `nests` is done at the end, rather than when encountered, which is required to produce results expected by the test suite.

### CODE REVIEW

File: lib/compact.js [hunk 1]
```diff
@@ -161,9 +160,9 @@ api.compact = ({
         } else {
           // expanded value must be a @type array
           compactedValue = [];
-          for(let vi = 0; vi < expandedValue.length; ++vi) {
+          for(let expandedIri of expandedValue) {
```
Comment (dlongley) [inline lib/compact.js]: Looks like could use `const expandedIri` instead of `let`.

File: lib/compact.js [hunk 2]
```diff
@@ -257,14 +256,21 @@ api.compact = ({
...
         _addValue(
-          rval, itemActiveProperty, expandedValue, {propertyIsArray: true});
+          nestResult, itemActiveProperty, expandedValue, {propertyIsArray: true});
       }
 
       // recusively process array values
-      for(let vi = 0; vi < expandedValue.length; ++vi) {
-        const expandedItem = expandedValue[vi];
-
+      for(let expandedItem of expandedValue) {
```
Comment (dlongley) [inline lib/compact.js]: Use `const` if `expandedItem` is not set anywhere in the loop.

File: lib/compact.js [hunk 3]
```diff
@@ -143,8 +143,7 @@ api.compact = ({
 
     // process element keys in order
     const keys = Object.keys(element).sort();
-    for(let ki = 0; ki < keys.length; ++ki) {
-      const expandedProperty = keys[ki];
+    for(let expandedProperty of keys) {
```
Comment (dlongley) [inline lib/compact.js]: Use `const` if `expandedProperty` is not set anywhere in the loop.

File: lib/compact.js [hunk 4]
```diff
@@ -906,3 +924,18 @@ function _selectTerm(
...
+ * @param activeCtx the active context.
+ * @param nestProperty a term in the active context or `@nest`.
+ */
+function _checkNestProperty(activeCtx, nestProperty) {
+  if(_expandIri(activeCtx, nestProperty, {vocab: true}) !== '@nest') {
+    throw new JsonLdError(
+      'JSON-LD compact error; nested property must have an @nest value resolving to @nest.',
+      'jsonld.SyntaxError', {code: 'invalid @nest value'});
+  }
+}
```
Comment (dlongley) [inline lib/compact.js]: Add EOL

File: lib/context.js [hunk 5]
```diff
@@ -478,6 +484,17 @@ api.createTermDefinition = (activeCtx, localCtx, term, defined) => {
     mapping['@language'] = language;
   }
 
+  if('@nest' in value) {
+    let nest = value['@nest'];
```
Comment (dlongley) [inline lib/context.js]: `const nest`

File: lib/expand.js [hunk 6]
```diff
@@ -145,9 +145,8 @@ api.expand = ({
       {activeCtx, localCtx: element['@context'], options});
   }
 
-  let keys = Object.keys(element).sort();
-
   // look for scoped context on @type
+  let keys = Object.keys(element).sort();
```
Comment (dlongley) [inline lib/expand.js]: `const keys`

File: lib/expand.js [hunk 7]
```diff
@@ -162,11 +161,172 @@ api.expand = ({
     }
   }
 
+
```
Comment (dlongley) [inline lib/expand.js]: Unnecessary added whitespace

File: lib/expand.js [hunk 8]
```diff
@@ -475,130 +640,32 @@ api.expand = ({
...
-        insideList
-      });
-      if(mapped !== undefined) {
-        rval = mapped;
-      } else {
-        rval = null;
+  // expand each nested key
+  for(let key of nests) {
+    const nestedValues = _isArray(element[key]) ? element[key] : [element[key]];
+    for(let nv of nestedValues) {
```
Comment (dlongley) [inline lib/expand.js]: `const`?

File: lib/expand.js [hunk 9]
```diff
@@ -475,130 +640,32 @@ api.expand = ({
...
-        element,
-        options,
-        insideList
-      });
-      if(mapped !== undefined) {
-        rval = mapped;
-      } else {
-        rval = null;
+  // expand each nested key
+  for(let key of nests) {
```
Comment (dlongley) [inline lib/expand.js]: `const`?

**Comment by gkellogg** [comment]
Those two are actually variables, as they're modified inline. This is questionable style, but for whatever reason, I chose not to create new variable to hold a possibly modified version.

**Comment by dlongley** [comment]
@gkellogg,

I see that the `keys` one is modified, but the `nest` one is not, AFAICT:

https://github.com/digitalbazaar/jsonld.js/blob/7baabf86f840c92091e23eacdef2513c1c08932d/lib/context.js#L487-L496

**Comment by dlongley** [comment]
@gkellogg, also please make sure to use rebase rather than merge to bring in other updates from 0.6.x, it keeps the history cleaner.

**Comment by gkellogg** [comment]
I've been re-basing, but the GitHub UI for resolving conflicts seems to have done a merge, I can probably re-do on the desktop.

**Comment by gkellogg** [comment]
Force pushed rebase.

**Pull request merged**